### PR TITLE
Sync between Prometheus Alertmanager's & Faythe's silences

### PR DESCRIPTION
**Pull request by ntk148v**
Closes-Issue: #34

### CODE REVIEW

File: pkg/cluster/cluster.go [hunk 1]
```diff
@@ -40,7 +40,7 @@ type State int
 
 const (
 	// DefaultLeaseTTL etcd lease time-to-live in seconds
-	DefaultLeaseTTL int64 = 10
+	DefaultLeaseTTL int64 = 20
```
Comment (vtdat) [inline pkg/cluster/cluster.go]: is this intended or what?

File: pkg/model/silence.go [hunk 2]
```diff
@@ -45,23 +47,28 @@ func (s *Silence) Validate() error {
...
 	if err != nil {
 		return err
 	}
 	s.RegexPattern = regex
 
-	t, err := common.ParseDuration(s.TTL)
-	if err != nil {
-		return err
+	if s.TTL == "" && !s.CreatedAt.IsZero() && !s.ExpiredAt.IsZero() {
+		durationHours := int(math.RoundToEven(s.ExpiredAt.Sub(s.CreatedAt).Hours()))
```
Comment (vtdat) [inline pkg/model/silence.go:57]: why only `hours` supported?

File: pkg/metrics/backends/prometheus/prometheus.go [hunk 3]
```diff
@@ -93,3 +97,51 @@ func (b *Backend) QueryInstant(ctx context.Context, query string, ts time.Time)
 		return nil, errors.Errorf("unknown supported type: '%q'", v)
 	}
 }
+
+// getAlertmanagers returns Alertmanager clients that is associated with the backend.
+func (b *Backend) getAlertmanagers(ctx context.Context) ([]*amclient.Alertmanager, error) {
+	level.Debug(b.logger).Log("msg", "Get active Alertmanagers")
+	amResult, err := b.prometheus.AlertManagers(ctx)
```
Comment (vtdat) [inline pkg/metrics/backends/prometheus/prometheus.go:104]: what if there is no Alertmanager configured?

File: pkg/metrics/backends/prometheus/prometheus.go [hunk 4]
```diff
@@ -93,3 +97,51 @@ func (b *Backend) QueryInstant(ctx context.Context, query string, ts time.Time)
...
+	if err != nil {
+		return nil, err
+	}
+	var ams []*amclient.Alertmanager
+	// Get only active Alertmanager instances.
+	for _, a := range amResult.Active {
+		u, _ := url.Parse(a.URL)
```
Comment (vtdat) [inline pkg/metrics/backends/prometheus/prometheus.go]: Should handle errors here 'cuz Alertmanager Url can be misconfigured in Prometheus?

File: pkg/metrics/backends/prometheus/prometheus.go [hunk 5]
```diff
@@ -93,3 +97,51 @@ func (b *Backend) QueryInstant(ctx context.Context, query string, ts time.Time)
...
+			Schemes:  amclient.DefaultSchemes,
+		})
+		ams = append(ams, amCli)
+	}
+	return ams, nil
+}
+
+// GetAlertmanagerSilences returns silences in Alertmanagers.
+func (b *Backend) GetAlertManagerSilences(ctx context.Context, filter []string) (map[string]ammodels.Silence, error) {
+	level.Debug(b.logger).Log("msg", "Get Alertmanager silences")
```
Comment (vtdat) [inline pkg/metrics/backends/prometheus/prometheus.go:131]: Debug log are overloaded. throw it at https://github.com/vCloud-DFTBA/faythe/pull/35/files#diff-c8a3dc6adf2ccc913d7cc34f8165bac5f8757f5924628e23da834e404ac4937eR143 to notify valid silences.

File: pkg/metrics/backends/prometheus/prometheus.go [hunk 6]
```diff
@@ -93,3 +97,51 @@ func (b *Backend) QueryInstant(ctx context.Context, query string, ts time.Time)
 		return nil, errors.Errorf("unknown supported type: '%q'", v)
 	}
 }
+
+// getAlertmanagers returns Alertmanager clients that is associated with the backend.
+func (b *Backend) getAlertmanagers(ctx context.Context) ([]*amclient.Alertmanager, error) {
```
Comment (vtdat) [inline pkg/metrics/backends/prometheus/prometheus.go:102]: It's quite wasteful to discover Alertmanagers every single healer queries, is there any ways we can cache or in-mem stored those AMs? It does not change frequently.

File: pkg/autohealer/healer.go [hunk 7]
```diff
@@ -428,3 +432,74 @@ func (h *Healer) updateSilence(e *common.Etcd) {
 		h.silences[sid] = &s
 	}
 }
+
+// syncSilencesFromBackend queries silences from the metric backend. Only Prometheus Alertmanager
+// is supported at this time.
+func (h *Healer) syncSilencesFromBackend(ctx context.Context, e *common.Etcd) {
+	if h.backend.GetType() != model.PrometheusType || !h.SyncSilences {
+		level.Info(h.logger).Log("msg", "Skip sync silences",
```
Comment (vtdat) [inline pkg/autohealer/healer.go]: Move to debug or just omit it. Just notify when users create healer that their monitor backend does not support sync silences.

File: pkg/autohealer/healer.go [hunk 8]
```diff
@@ -428,3 +432,74 @@ func (h *Healer) updateSilence(e *common.Etcd) {
...
+			"backend", h.backend.GetAddress())
+		return
+	}
+	level.Info(h.logger).Log("msg", "Sync silences from metric backend",
```
Comment (vtdat) [inline pkg/autohealer/healer.go]: Debug please

File: pkg/autohealer/healer.go [hunk 9]
```diff
@@ -428,3 +432,74 @@ func (h *Healer) updateSilence(e *common.Etcd) {
...
+			if *matcher.Name == "instance" {
+				s.Pattern = *matcher.Value
+				break
+			}
+		}
+		if s.Pattern == "" {
+			continue
+		}
+		if err := s.Validate(); err != nil {
+			level.Error(h.logger).Log("msg", "Error when validating silence", "err", err)
```
Comment (vtdat) [inline pkg/autohealer/healer.go]: is there any cases that silences is valid in AM and not valid here? that will throw a lot of errors.

**Review by vtdat — COMMENTED** [review]
Some comments and questions

File: pkg/model/silence.go (hunk 2 above)
Comment (ntk148v) [reply pkg/model/silence.go:57]: `common.ParseDuration` doesn't support parsing string like `10d2h`. Therefore, we have to convert it to hours then round up.

File: pkg/metrics/backends/prometheus/prometheus.go (hunk 3 above)
Comment (ntk148v) [reply pkg/metrics/backends/prometheus/prometheus.go:104]: It's fine, amResult will be an empty slice.

File: pkg/metrics/backends/prometheus/prometheus.go (hunk 4 above)
Comment (ntk148v) [reply pkg/metrics/backends/prometheus/prometheus.go]: Ok, actually if there is an invalid URL, a 503 status code error will be returned when we use the client.

File: pkg/metrics/backends/prometheus/prometheus.go (hunk 6 above)
Comment (ntk148v) [reply pkg/metrics/backends/prometheus/prometheus.go:102]: I used a separate ticker loop in the first place.

File: pkg/autohealer/healer.go (hunk 9 above)
Comment (ntk148v) [reply pkg/autohealer/healer.go]: Not much, after TTL logic is updated. The Validate call is for generating ID and TTL.

File: pkg/model/const.go [hunk 10]
```diff
@@ -34,6 +34,7 @@ const (
 	// DefaultSilencePrefix is default etcd prefix for Silences
 	DefaultSilencePrefix             = "/silences"
 	DefaultSilenceValidationInterval = "30s"
+	DefaultSyncSilencesInterval      = "60s"
```
Comment (vtdat) [inline pkg/model/const.go:37]: Would love to have that documented :D

Comment (vtdat) [reply pkg/model/const.go:37]: So that users can be aware that their silences might take up to 60 seconds to be syned.

**Review by vtdat — APPROVED** [review]
lgtm

**Pull request merged**